packages/labs/context/README.md

```diff
@@ -0,0 +1,101 @@
+# @lit-labs/context
+
+## Overview
+
+## Usage
+
+There are several different usages of the Context API.
+
+### Creating a Context
+
+First lets define a context key we can use elsewhere in our examples:
+
+```ts
+import {createContext} from '@lit-labs/context';
+
+export interface Logger {
+  log: (msg: string) => void;
+}
+
+export const loggerContext = createContext<Logger>('logger');
+```
+
+### Consuming a Context
+
+Now we can define a consumer for this context, some component in our app needs the logger:
+
+```ts
+import {ContextConsumer, property} from '@lit-labs/context';
+import {LitElement} from 'lit';
+import {Logger, loggerContext} from './logger.js';
+
+export class MyElement extends LitElement {
+  @property({attribute: false})
+  public logger?: Logger;
+
+  public constructor() {
+    // add the consumer controller
+    this.addController(
```

Comment:
ContextConsumer already adds itself as a controller, so the `addController()` call isn't necessary.

Also, the patterns are new with controllers, but I think the pattern we're gravitating to is that controller contains its state and requests a host update when that state changes, then other code can just access it:

```ts
export class MyElement extends LitElement {
  private _loggerContext = new ContextConsumer(this, loggerContext);
  
  render() {
    this._loggerContext.value?.log('rendering...');
    return html`...`;
  }
}
```

In cases where code needs to know that _specifically_ this context was updated, the controller can take an optional callback.

Reply:
That pattern is an interesting one. I certainly agree with it where a context *has* to be provided, but if the context is designed as an optional pattern such that it can be used to fulfill a property value then I guess its not quite as straightforward.

For example in the `@contextProvided` decorator case, we're using the ContextConsumer controller to populate the property, a user of the component can just directly bind a property value or provide it via context. This seems very flexible to me.

Reply:
I think in that case an element could use the callback, though the author should be very careful about how two different sources for a value interact. Most likely they want the local property to win, so either the callback should use logical nullish assignment, or a getter with a fallback to the ContextConsumer value:

```ts
class MyElement extends LitElement {
  private _loggerContext = new ContextConsumer(this, loggerContext);
  
  private _logger?: Logger;
  get logger() {
    return this._logger ?? this._loggerContext.value;
  }
  set logger(v: Logger) {
    this._logger = v;
    this.requestUpdate();
  }
}
```

I think it'd be good if `@contextProvided` allowed this. Right now it looks like it always overwrites the local value.

Reply:
I'm a little unsure how to go about this. Shoudl the `@contextProvided` directive add the property with getter/setter to the element if it doesn't already have a property? Is there any bad interaction that might have for `@property` decorator? Or should we just to the logical nullish assignment on receipt of a new context value instead?

Reply:
Was there something here still remaining @justinfagnani ?

Reply:
After offline discussion marking this as resolved after all the changes we've made since this thread.